### PR TITLE
Farley exercise 5

### PR DESCRIPTION
**Pull request by ivanakrtolicaRTRK**

### CODE REVIEW

**Review by momokrunic — COMMENTED** [review]
# 🧩 **PR Review — Team Farley (Exercise 5)**

## 🎯 **Overview**

This PR completes **Exercise 5**, introducing a structured and fully functional **acceptance-testing layer** that integrates all system components (`Filter`, `ThermalReader`, `RawTempFacade`, and the C simulation library).
 It implements parameterized tests, updates the build system, and cleanly separates unit and acceptance test layers.
 Technically solid — main refinements should focus on **naming consistency** and **test-layer cohesion**.

------

## 🗂 **1 · Project Structure**

- ✅ Introduces a dedicated **`test/acceptance/`** directory for high-level, end-to-end tests.
- ✅ Retains **`test/unit/`** for low-level component verification.
- ✅ Each layer has its own `CMakeLists.txt`, ensuring acceptance tests link **real components** and **exclude mocks**.
- ✅ `inc/`, `src/`, and `mock/` are properly scoped and cleanly organized.
- ⚠️ **Cohesion improvement:**
  - Keep **acceptance tests** exclusively in `test/acceptance/`.
  - Keep **unit tests** exclusively in `test/unit/`.
  - Avoid overlapping or hybrid files mixing mocks and real objects.

**Verdict:** Excellent modularity — ensure each layer owns only its relevant tests.

------

## ⚙️ **2 · Use of Real Implementations**

- ✅ Acceptance tests instantiate **real** production classes — no mocks or stubs.
- ✅ Full integration path verified: `HardwareMonitor → ThermalReader → RawTempFacade → external C library`.
- ✅ Acceptance builds exclude `MOCK_ENABLE`, guaranteeing production-like behavior.

**Verdict:** Perfectly aligned with Exercise 5’s goal — test the integrated, real system.

------

## 🧩 **3 · Parameterized Testing (TEST_P)**

- ✅ Proper use of **GoogleTest `TEST_P`** and `INSTANTIATE_TEST_SUITE_P`.
- ✅ Parameterized structs (e.g., `{ numOfUpdates, expectedFilteredValue }`) allow concise, reusable coverage.
- ✅ Logic correctly loops over `UpdateCurrentTemp()` calls and checks expected results.

**Verdict:** Clean, efficient, and professional parameterized test design.

------

## 🌡 **4 · C Library Integration**

- ✅ Acceptance tests link directly to the real **`external_c_lib.c`** implementation.
- ✅ The simulated temperature sequence (0 → 120 → reset) is processed through all system layers.
- ✅ Tests verify correct data propagation and filtering.
- ⚠️ **Edge-case suggestion:** Add one parameterized case covering the **wrap-around point (120 → 0)** to confirm correct rollover handling.

**Verdict:** Realistic hardware simulation achieved and well validated.

------

## 🧱 **5 · Code Quality & Naming Consistency**

- ✅ Code and comments clearly explain purpose and expectations.
- ✅ File hierarchy mirrors professional C++ layout.
- ⚠️ **Test-name unification required:**
   Current files mix styles (`at_hardware_monitor.cpp`, `test_hardware_monitor.cpp`, `ut_hardware_monitor.cpp`).
   Adopt a unified pattern to make the test intent immediately clear:
  - Prefix **acceptance tests** with `AT_` (e.g., `AT_HardwareMonitor_WhenTemperatureUpdated_ThenFilterReceivesNewValue`).
  - Prefix **unit tests** with `UT_`.
  - Use consistent fixture naming (`HardwareMonitorAcceptance` vs `HardwareMonitorTests`).
  - Apply a behavior-driven pattern — `When_<Action>_Then_<ExpectedResult>`.

**Verdict:** Functionally excellent — naming and structure harmonization will strengthen clarity and cohesion.

------

## 💡 **Additional Suggestions**

1. **Reinforce Layer Boundaries**
   - Make clear code cohesion by separating unit and acceptance test in appropriate folders.
2. **Improve Tests Naming Convention**
   * There is a mixture of test naming conventions and their folder placement: `at_*`, `test_*`, `ut_*`. 
3. **Add Reset Behavior Test**
   - Cover the 120 → 0 sensor-reset scenario to confirm correct filtering continuity.

------

## ✅ **Final Assessment**

| Criterion             | Evaluation                                  |
| --------------------- | ------------------------------------------- |
| Project Structure     | ✅ Clear, layered, and modular               |
| Real Implementations  | ✅ Integrated and verified                   |
| Parameterized Tests   | ✅ Efficient and readable                    |
| C Library Integration | ✅ Realistic, minor edge case suggested      |
| Naming & Cohesion     | ⚠️ Unify prefixes and directory ownership    |
| Overall Readiness     | ✅ Merge-ready after naming and cohesion fix |

------

### 🟩 **Conclusion**

Team Farley’s PR fully meets the technical and pedagogical objectives of **Exercise 5**:

- Structured acceptance-test environment,
- End-to-end testing of real implementations,
- Parameterized test coverage for realistic data flows,
- Proper integration with the C hardware simulation.

File: UserStories.md [hunk 1]
```diff
@@ -158,4 +158,97 @@ I want to configure MockFilter to return different values on successive and repe
...
+**And** log the event with timestamp and temperature value.
+ 
+**Scenario 3: Underheating Detection and Alarm Reporting**
+ 
+**Given** the filtered temperature falls below the defined minimum threshold
+**And** the temperature reading is plausible
+**When** the HardwareMonitor compares the temperature to the thresholds
+**Then** it should report an underheating alarm to the SystemAlarmHandler
+**And** log the event with timestamp and temperature value.
+
```
Comment (momokrunic) [inline UserStories.md:237]: # 🧩 **User Story Review — Team Farley (Exercise 5)**

## 🎯 **Overview**

This set of user stories (5.1 – 5.3) defines and documents the transition from isolated unit tests to full acceptance testing using **real system components** and **parameterized scenarios**.
 Additionally, the stories for the **Hardware Monitor** project extend the acceptance layer by validating continuous temperature monitoring and alarm reporting under realistic conditions.
 Overall, these stories demonstrate a strong understanding of system-level testing, maintainability, and test architecture.

------

## 🗂 **User Story 5.1 — Separate Acceptance Tests from Unit Tests**

**Evaluation:**
 ✅ Clearly defines the need to separate **acceptance** and **unit** test layers.
 ✅ The `/test/acceptance` directory and its dedicated CMake configuration fulfill the structural and architectural requirement.
 ✅ Scenarios 1 and 2 explicitly describe both the organizational and build aspects — ensuring that real implementations are compiled and linked cleanly.

**Highlights:**

- The distinction between **unit** and **acceptance** tests is explicit and well executed.
- Proper linking of real dependencies (`raw_temp` C library, Filter, ThermalReader) ensures modularity and prevents cross-contamination with mocks.
- CMake integration matches real-world industry practice for multi-layer testing setups.

**Verdict:**
 ✔ Meets all acceptance-test separation and build-integration criteria.
 ⚙️ Recommendation: reinforce this in documentation (`README` or `UserStories.md`) by summarizing directory purpose and test scope.

------

## 🧩 **User Story 5.2 — Use Real Implementations in Acceptance Tests**

**Evaluation:**
 ✅ Acceptance tests instantiate **real classes only** — no mocks or stubs are included.
 ✅ Test execution validates real data flow across `RawTempFacade`, `ThermalReader`, and `Filter`.
 ✅ The story explicitly captures the goal of **verifying the integrated system’s behavior under realistic conditions**.

**Highlights:**

- Tests correctly use the real `external_c_lib.c`, ensuring that thermal-reader behavior reflects live-data flow.
- System integration mirrors production operation, validating the correctness of DI (dependency injection).
- This story effectively bridges the gap between unit verification and integration validation.

**Verdict:**
 ✔ Fully accomplished — implementation and documentation are in perfect alignment.
 💡 Consider noting any system assumptions (e.g., default thresholds or expected range) in acceptance-test comments for clarity.

------

## 🧠 **User Story 5.3 — Parameterized Acceptance Tests with Real C Library Input**

**Evaluation:**
 ✅ Uses **GoogleTest TEST_P and INSTANTIATE_TEST_SUITE_P** patterns correctly.
 ✅ Parameter structures capture both iteration counts and expected results, avoiding redundant code.
 ✅ Real C library input drives test realism — verifying looped temperature progression (0 → 120 → reset).

**Highlights:**

- Story scenarios correspond directly to implementation in `test_thermal_reader.cpp` and `test_hardware_monitor.cpp`.
- Parameterized design allows easy extension for future edge-case coverage.
- Demonstrates good balance between realism and maintainability.

**Verdict:**
 ✔ Fully satisfies parameterization goals; test logic is clean and scalable.
 ⚙️ Recommendation: add one edge parameter set covering the wrap-around point (120 → 0) to guarantee robustness.

------

## 🌡 **Hardware Monitor Project Stories (Temperature and Alarm Logic)**

**Evaluation:**
 ✅ The hardware-monitor scenarios correctly extend acceptance-testing principles into a higher-level integration layer.
 ✅ The stories focus on continuous operation, threshold comparison, and alarm triggering — all realistic embedded-system concerns.
 ✅ Scenarios 1–3 clearly express *cause → action → effect* relationships suitable for future BDD automation.

**Highlights:**

- Describes logical sequencing (sampling → filter → compare → alarm).
- Reinforces modular responsibility — `HardwareMonitor` aggregates `Codings` and `SystemAlarmHandler`.
- Complements Exercise 5 acceptance tests by providing the functional base for Exercise 6.

**Verdict:**
 ✔ Conceptually solid and technically feasible; ready for direct implementation in future iterations.
 💡 Minor suggestion: document expected numeric thresholds (min/max) within `Codings` tests for reference clarity.

------

## ⚙️ **Summary of Review**

| **Criterion**                           | **Evaluation**                                   |
| --------------------------------------- | ------------------------------------------------ |
| Separation of Acceptance and Unit Tests | ✅ Implemented and well structured                |
| Real Implementation Usage               | ✅ Fully achieved                                 |
| Parameterized Testing Coverage          | ✅ Efficient and clear                            |
| Hardware Monitor Integration Scenarios  | ✅ Comprehensive and logical                      |
| Documentation Clarity                   | ✅ Excellent alignment between story and code     |
| Naming & Consistency                    | ⚙️ Unified prefix (AT_/UT_) recommended for tests |

------

## 🟩 **Conclusion**

Team Farley’s User Stories for Exercise 5 demonstrate a mature, professional understanding of how to evolve from isolated unit testing to fully integrated acceptance testing.
 They establish clear structure, realistic simulation, and parameterized test design — all essential foundations for system-level TDD and BDD workflows.

> ✅ **Final Assessment:**
>
> - Architecture and user stories are coherent and complete.
> - Implementation directly satisfies every defined goal.
> - Next step: ensure **consistent naming (AT_/UT_)** and maintain cohesion between test layers in future PRs.

**Comment by momokrunic** [comment]
@SandraIvanovic please add submission to the Canvas:
https://canvas.nit-institute.com/courses/121/gradebook/speed_grader?assignment_id=2078&student_id=715

@ivanakrtolicaRTRK please add submission to the Canvas:
https://canvas.nit-institute.com/courses/121/gradebook/speed_grader?assignment_id=2078&student_id=712
